File: CFOnlineShop/CFOnlineShop/ViewController/CFHomePageController/CFDetailViewController.m
//
//  CFDetailViewController.m
//  CFOnlineShop
//
//  Created by chenfeng on 2018/8/6.
//  Copyright © 2018年 chenfeng. All rights reserved.
//

#import "CFDetailViewController.h"
//#import "CFDetailView.h"

@interface CFDetailViewController ()<UITableViewDelegate,UITableViewDataSource>

@property (nonatomic, strong) UIView *bigView;
@property (nonatomic, strong) UIWebView *webView;

//添加在头部视图的tempScrollView
@property (nonatomic, strong) UIScrollView *tempScrollView;
//记录底部空间所需的高度
@property (nonatomic, assign) CGFloat bottomHeight;

//@property (nonatomic, strong) CFDetailView *detailView;

@end

@implementation CFDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self setBgUI];
    [self setHeaderAndFooterView];
    [self setBottomView];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)dealloc
{
    _webView.delegate = nil;
    _webView = nil;
    NSLog(@"释放View");
}

- (void)setBgUI
{
    _bottomHeight = TabbarHeight;
    
    //存放tableView和webView，tableview在上面，webview在下面
    _bigView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, (Main_Screen_Height - _bottomHeight) * 2)];
    _bigView.backgroundColor = kWhiteColor;
    
    _tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Height - _bottomHeight)];
    _tableView.backgroundColor = kWhiteColor;
    _tableView.delegate = self;
    _tableView.dataSource = self;
    
    //去掉顶部偏移
    if (@available(iOS 11.0, *))
    {
        _tableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    }
    
    _webView = [[UIWebView alloc] initWithFrame:CGRectMake(0, Main_Screen_Height - _bottomHeight, Main_Screen_Width, Main_Screen_Height - _bottomHeight)];
    _webView.backgroundColor = [UIColor clearColor];
    
    _webView.scrollView.delegate = self;
    
    [self.view addSubview:_bigView];
    [_bigView addSubview:_tableView];
    [_bigView addSubview:_webView];
    
    WeakSelf(self);
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakself.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://www.baidu.com"]]];
    });
}

- (void)setHeaderAndFooterView{
    
    //添加头部和尾部视图
    UIView *headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Width)];
    headerView.backgroundColor = kWhiteColor;
    
    _tempScrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Width)];
    [headerView addSubview:_tempScrollView];
    
    _headerImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, Main_Screen_Width)];
    _headerImageView.contentMode = UIViewContentModeScaleAspectFit;
    _headerImageView.image = _image;
    [_tempScrollView addSubview:_headerImageView];
    
    _tableView.tableHeaderView = headerView;
    
    UILabel *pullMsgView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, 40)];
    pullMsgView.textAlignment = NSTextAlignmentCenter;
    pullMsgView.text = @"上拉显示网页";
    pullMsgView.textColor = KGrayTextColor;
    
    UIView *footView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, 40)];
    [footView addSubview:pullMsgView];
    
    _tableView.tableFooterView = footView;
    
    //设置下拉提示视图
    UILabel *downPullMsgView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, 40)];
    downPullMsgView.textAlignment = NSTextAlignmentCenter;
    downPullMsgView.text = @"下拉显示列表";
    downPullMsgView.textColor = KGrayTextColor;
    
    UIView *downMsgView = [[UIView alloc] initWithFrame:CGRectMake(0, -40, Main_Screen_Width, 40)];
    [downMsgView addSubview:downPullMsgView];
    [_webView.scrollView addSubview:downMsgView];
}

- (void)setBottomView
{
    UIView *bottomView = [[UIView alloc] initWithFrame:CGRectMake(0, Main_Screen_Height - _bottomHeight, Main_Screen_Width, _bottomHeight)];
    bottomView.backgroundColor = KBackgroundColor;
    [self.view addSubview:bottomView];
    
    UIButton *addButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
    addButton.frame = CGRectMake(bottomView.mj_w/2, 0, bottomView.mj_w/2, _bottomHeight);
    addButton.backgroundColor = kRedColor;
    addButton.titleLabel.font = SYSTEMFONT(16);
    [addButton setTitle:@"加入购物车" forState:(UIControlStateNormal)];
    [addButton setTitleColor:kWhiteColor forState:(UIControlStateNormal)];
    [addButton addTarget:self action:@selector(addAction) forControlEvents:(UIControlEventTouchUpInside)];
    [bottomView addSubview:addButton];
    
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGFloat offset = scrollView.contentOffset.y;
    
    if (scrollView == _tableView){
        //重新赋值，就不会有用力拖拽时的回弹
        _tempScrollView.contentOffset = CGPointMake(_tempScrollView.contentOffset.x, 0);
        if (offset >= 0 && offset <= Main_Screen_Width) {
            //因为tempScrollView是放在tableView上的，tableView向上速度为1，实际上tempScrollView的速度也是1，此处往反方向走1/2的速度，相当于tableView还是正向在走1/2，这样就形成了视觉差！
            _tempScrollView.contentOffset = CGPointMake(_tempScrollView.contentOffset.x, - offset / 2.0f);
        }
    }
    
    _scrollViewDidScroll(scrollView);
}

#pragma mark -- 监听滚动实现商品详情与图文详情界面的切换
-(void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate{
    
    WeakSelf(self);
    CGFloat offset = scrollView.contentOffset.y;
    if (scrollView == _tableView) {
        if (offset > _tableView.contentSize.height - Main_Screen_Height + self.bottomHeight + 50) {
            [UIView animateWithDuration:0.4 animations:^{
                weakself.bigView.transform = CGAffineTransformTranslate(weakself.bigView.transform, 0, -Main_Screen_Height +  self.bottomHeight + TopHeight);
            } completion:^(BOOL finished) {
                
            }];
        }
    }
    if (scrollView == _webView.scrollView) {
        if (offset < -50) {
            [UIView animateWithDuration:0.4 animations:^{
                [UIView animateWithDuration:0.4 animations:^{
                    weakself.bigView.transform = CGAffineTransformIdentity;
                    
                }];
            } completion:^(BOOL finished) {
                
            }];
        }
    }
}

#pragma mark - action

- (void)addAction
{
    if (_addActionWithBlock) {
        _addActionWithBlock();
    }
}

#pragma mark -- UITableViewDelegate & dataSource

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 10;
}


- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 50;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *identifier = @"identifier";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:identifier];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
    }
    
    cell.textLabel.font = SYSTEMFONT(16);
    cell.textLabel.textColor = KDarkTextColor;
    cell.textLabel.text = @"哈哈哈哈哈啊哈哈";
    
    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftOnlineShopDemo/SwiftOnlineShopDemo/ViewController/CFHomePageController/CFDetailViewController.swift
//
//  CFDetailViewController.swift
//  SwiftOnlineShopDemo
//
//  Created by chenfeng on 2019/4/26.
//  Copyright © 2019年 chenfeng. All rights reserved.
//

import UIKit
import WebKit

class CFDetailViewController: CFBaseController,UITableViewDelegate,UITableViewDataSource,CFCycleScrollViewDataSource,WKNavigationDelegate {
    
    var tableView: UITableView!
    //闭包-类似于oc中的block
    var addActionWithBlock: (() ->())?
    var scrollViewDidScroll_Closure: ((UIScrollView) ->())?
    var bigView: UIView!
    var webView: WKWebView!
    var tempScrollView: UIScrollView!
    var cycleView: CFCycleScrollView!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        setBgUI()
        setHeaderAndFooterView()
        setBottomView()
    }
    
    func setBgUI() {
        
        //存放tableView和webView，tableview在上面，webview在下面
        bigView = UIView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: (SCREEN_HEIGHT - TABBAR_HEIGHT) * 2))
        bigView.backgroundColor = UIColor.white
        
        tableView = UITableView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_HEIGHT - TABBAR_HEIGHT))
        tableView.backgroundColor = UIColor.white;
        tableView.delegate = self;
        tableView.dataSource = self;
        
        //去掉顶部偏移
        if #available(iOS 11.0, *)
        {
            tableView.contentInsetAdjustmentBehavior = .never;
        }
        
        webView = WKWebView(frame: CGRect(x: 0, y: SCREEN_HEIGHT - TABBAR_HEIGHT, width: SCREEN_WIDTH, height: SCREEN_HEIGHT - TABBAR_HEIGHT))
        webView.navigationDelegate = self
        webView.scrollView.delegate = self
        
        view.addSubview(bigView)
        bigView.addSubview(tableView)
        bigView.addSubview(webView)
        
        DispatchQueue.main.async {
            self.webView.load(URLRequest(url: URL(string: "https://www.baidu.com")!))
        }
    }
    
    func setHeaderAndFooterView() {
        //添加头部和尾部视图
        let headerView = UIView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_WIDTH))
        headerView.backgroundColor = UIColor.white
        
        tempScrollView = UIScrollView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_WIDTH))
        headerView.addSubview(tempScrollView)
        
        cycleView = CFCycleScrollView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: SCREEN_WIDTH))
        cycleView.dataSource = self;
        tempScrollView.addSubview(cycleView)
        
        tableView.tableHeaderView = headerView
        
        let pullMsgView = UILabel.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: 40))
        pullMsgView.textAlignment = .center
        pullMsgView.text = "上拉显示网页"
        pullMsgView.textColor = UIColor.gray
        
        let footView = UIView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: 40))
        footView.addSubview(pullMsgView)
        
        tableView.tableFooterView = footView
        
        //设置下拉提示视图
        let downPullMsgView = UILabel.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: 40))
        downPullMsgView.textAlignment = .center
        downPullMsgView.text = "下拉显示列表"
        downPullMsgView.textColor = UIColor.gray
        
        let downMsgView = UIView.init(frame: CGRect(x: 0, y: -40, width: SCREEN_WIDTH, height: 40))
        downMsgView.addSubview(downPullMsgView)
        webView.scrollView.addSubview(downMsgView)
    }
    
    func setBottomView() {
        
        let bottomView = UIView.init(frame: CGRect(x: 0, y: SCREEN_HEIGHT - TABBAR_HEIGHT, width: SCREEN_WIDTH, height: TABBAR_HEIGHT))
        bottomView.backgroundColor = BACKGROUND_VIEW_COLOR;
        self.view.addSubview(bottomView)
        
        let addButton = UIButton(type: .custom)
        addButton.frame = CGRect(x: bottomView.mj_w/2, y: 0, width: bottomView.mj_w/2, height: TABBAR_HEIGHT)
        addButton.backgroundColor = UIColor.red;
        addButton.titleLabel?.font = UIFont.systemFont(ofSize: 16);
        addButton.setTitle("加入购物车", for: .normal)
        addButton.setTitleColor(UIColor.white, for: .normal)
        addButton.addTarget(self, action: #selector(addAction), for: .touchUpInside)
        bottomView.addSubview(addButton)
    }
    
    @objc func addAction() {
        addActionWithBlock?()
    }
    
    // MARK: - UIScrollViewDelegate
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        
        let offset = scrollView.contentOffset.y;
        
        if scrollView == tableView{
            //重新赋值，就不会有用力拖拽时的回弹
            tempScrollView.contentOffset = CGPoint(x: tempScrollView.contentOffset.x, y: 0)
            if offset >= 0 && offset <= SCREEN_WIDTH {
                //因为tempScrollView是放在tableView上的，tableView向上速度为1，实际上tempScrollView的速度也是1，此处往反方向走1/2的速度，相当于tableView还是正向在走1/2，这样就形成了视觉差！
                tempScrollView.contentOffset = CGPoint(x: tempScrollView.contentOffset.x, y: -offset / 2)
            }
        }
        
        scrollViewDidScroll_Closure?(scrollView)
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        let offset = scrollView.contentOffset.y;
        if scrollView == tableView {
            if offset > tableView.contentSize.height - SCREEN_HEIGHT + TABBAR_HEIGHT + 50 {
                
                UIView.animate(withDuration: 0.4) {
                    self.bigView.transform = self.bigView.transform.translatedBy(x: 0, y: -SCREEN_HEIGHT + TABBAR_HEIGHT + STATUS_AND_NAV_BAR_HEIGHT)
                }
            }
        }
        if scrollView == webView.scrollView {
            if offset < -50 {
                UIView.animate(withDuration: 0.4) {
                    self.bigView.transform = self.bigView.transform.translatedBy(x: 0, y: SCREEN_HEIGHT - TABBAR_HEIGHT - STATUS_AND_NAV_BAR_HEIGHT)
                }
            }
        }
    }
    
    // MARK: - tableView
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 10
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCell(withIdentifier: "CellIdentifier")
        
        if cell == nil {
            cell = UITableViewCell.init(style: UITableViewCell.CellStyle.default, reuseIdentifier: "CellIdentifier")
        }
        
        cell?.textLabel?.text = "哈哈哈哈哈"
        return cell!
    }
    
    // MARK: - CFCycleScrollViewDataSource
    func getImagesWithArray() -> Array<Any> {
        return ["commodity_1","commodity_2","commodity_3","commodity_4","commodity_5"]
    }
    
    // MARK: - wkWebView
    // 页面开始加载时调用
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        print("开始加载...")
    }
    
    // 当内容开始返回时调用
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!){
        print("当内容开始返回...")
    }
    
    // 页面加载完成之后调用
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!){
        print("页面加载完成...")
    }
    
    // 页面加载失败时调用
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error){
        print("页面加载失败...")
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
